skills_ml/algorithms/occupation_classifiers/train.py

```diff
@@ -0,0 +1,131 @@
+
+import numpy as np
+import importlib
+
+
+def get_all_soc():
+    onet = build_onet()
+    occupations = onet.occupations
+    soc = []
+    for occ in occupations:
+        if 'O*NET-SOC Occupation' in occ.other_attributes['categories']:
+            soc.append(occ.identifier)
+
+    return soc
+
+class OccupationClassifierTrainer(object):
+    """Trains a series of classifiers using the same training set
+    Args:
+        matrix ()
+        storage (skills_ml.storage)
+        target_variable (str)
+        embedding_model (gensim.model)
+        k_folds (int)
+        n_jobs (int)
+    """
+    def __init__(self, matrix, k_folds, grid_config=None, storage=None,
+                 random_state_for_split=None, scores=['accuracy'], n_jobs=3):
+        self.matrix = matrix
+        self.storage = FSStore() if storage is None else storage
+        self.target_variable = self.matrix['target_variable']
+        self.k_folds = k_folds
+        self.n_jobs = n_jobs
+        self.grid_config = self.default_grid_config if grid_config is None else grid_config
+        self.cls_cv_result = {}
+        self.scores = scores
+        self.best_classifiers = {}
+        self.random_state_for_split = random_state_for_split
+
+    @property
+    def default_grid_config(self):
+        return {
+                'sklearn.ensemble.ExtraTreesClassifier': {
+                    'n_estimators': [100, 500],
+                    'criterion': ['gini', 'entropy'],
+                    'max_depth': [1, 5, 10, 20, 50],
+                    'max_features': ['sqrt', 'log2'],
+                    'min_samples_split': [2, 5, 10]
+                    },
+                'sklearn.ensemble.RandomForestClassifier': {
+                    'n_estimators': [100, 500],
+                    'criterion': ['gini', 'entropy'],
+                    'max_depth': [1, 5, 10, 20, 50],
+                    'max_features': ['sqrt', 'log2'],
+                    'min_samples_split': [2, 5, 10]
+                    }
+                }
+
+    def train(self):
+        """Fit a model to a training set. Works on any modeling class that
+        is vailable in this package's environment and implements .fit
+        """
+        X = self.matrix['X']
+        y = self.matrix['y']
+        for score in self.scores:
+            for class_path, parameter_config in self.grid_config.items():
+                module_name, class_name = class_path.rsplit(".", 1)
+                module = importlib.import_module(module_name)
+                cls = getattr(module, class_name)
+                kf = StratifiedKFold(n_splits=self.k_folds, random_state=self.random_state_for_split)
+
+                cls_cv = GridSearchCV(cls(), parameter_config, cv=kf, scoring=score, n_jobs=self.n_jobs)
+                cls_cv.fit(X, y)
+                self.cls_cv_result[score] = {class_name: cls_cv.cv_results_}
+                self.best_classifiers[score] = {class_name: cls_cv}
+
+    def save_result(self):
```

Comment:
What is the plan for saving the classifiers we want to use? We have a storage object all ready to go and what looks like some classifiers to save.

Reply:
I don't think we're gonna save the model/classifier itself but the cross-validation results and model hyper-parameters which is going to be in another PR. It's gonna be like what we've been talking about that in the json format.

Reply:
I do remember talk about this, but maybe I misunderstood.

Ultimately we'll want to use the winner/s for prediction on a large dataset. I don't want to make the code that does this crawl through job postings again.

Is it that we will save the training matrix, so that usage just involves loading that matrix, the hyperparameters, and calling .fit?

Reply:
Ah right. sorry I made it confusing. It's not that we don't want to save the winner classifier but maybe not in this stage. The winner classifier in this stage is the winner in training/validation. It might be over-fitting to become the winner, so there will be a testing stage to see how it performs on our test dataset(e.g. our uniform 24k sample). I'm just not sure if it's the best thing to do to save the classifier right after the training stage.  The other question would be if we want to save classifier do we want to save all or just the winner in this stage.